File: mifugo-care/app/src/main/java/com/mifugocare/app/database/entities/Diagnosis.java
package com.mifugocare.app.database.entities;

import androidx.room.Entity;
import androidx.room.PrimaryKey;
import androidx.room.ForeignKey;
import androidx.room.Index;

@Entity(tableName = "diagnoses",
        foreignKeys = {
            @ForeignKey(entity = Livestock.class,
                       parentColumns = "livestockId",
                       childColumns = "livestockId",
                       onDelete = ForeignKey.CASCADE),
            @ForeignKey(entity = Disease.class,
                       parentColumns = "diseaseId",
                       childColumns = "diseaseId",
                       onDelete = ForeignKey.CASCADE)
        },
        indices = {@Index("livestockId"), @Index("diseaseId")})
public class Diagnosis {
    @PrimaryKey(autoGenerate = true)
    public int diagnosisId;
    
    public int livestockId;
    public int diseaseId;
    public String imagePath; // Path to the captured image
    public String diagnosisResult; // ML model prediction
    public float confidence; // Confidence score (0.0 - 1.0)
    public String status; // "Confirmed", "Suspected", "Negative"
    public String notes;
    public String veterinarianNotes;
    public String treatmentPrescribed;
    public String followUpDate;
    public boolean isTreated;
    public String location;
    public String weatherConditions;
    public long diagnosisDate;
    public long lastUpdated;
    
    // Constructors
    public Diagnosis() {
        this.diagnosisDate = System.currentTimeMillis();
        this.lastUpdated = System.currentTimeMillis();
        this.isTreated = false;
    }
    
    public Diagnosis(int livestockId, String imagePath, String diagnosisResult, 
                    float confidence, String status) {
        this();
        this.livestockId = livestockId;
        this.imagePath = imagePath;
        this.diagnosisResult = diagnosisResult;
        this.confidence = confidence;
        this.status = status;
    }
    
    // Getters and Setters
    public int getDiagnosisId() { return diagnosisId; }
    public void setDiagnosisId(int diagnosisId) { this.diagnosisId = diagnosisId; }
    
    public int getLivestockId() { return livestockId; }
    public void setLivestockId(int livestockId) { this.livestockId = livestockId; }
    
    public int getDiseaseId() { return diseaseId; }
    public void setDiseaseId(int diseaseId) { this.diseaseId = diseaseId; }
    
    public String getImagePath() { return imagePath; }
    public void setImagePath(String imagePath) { this.imagePath = imagePath; }
    
    public String getDiagnosisResult() { return diagnosisResult; }
    public void setDiagnosisResult(String diagnosisResult) { this.diagnosisResult = diagnosisResult; }
    
    public float getConfidence() { return confidence; }
    public void setConfidence(float confidence) { this.confidence = confidence; }
    
    public String getStatus() { return status; }
    public void setStatus(String status) { this.status = status; }
    
    public String getNotes() { return notes; }
    public void setNotes(String notes) { this.notes = notes; }
    
    public String getVeterinarianNotes() { return veterinarianNotes; }
    public void setVeterinarianNotes(String veterinarianNotes) { this.veterinarianNotes = veterinarianNotes; }
    
    public String getTreatmentPrescribed() { return treatmentPrescribed; }
    public void setTreatmentPrescribed(String treatmentPrescribed) { this.treatmentPrescribed = treatmentPrescribed; }
    
    public String getFollowUpDate() { return followUpDate; }
    public void setFollowUpDate(String followUpDate) { this.followUpDate = followUpDate; }
    
    public boolean isTreated() { return isTreated; }
    public void setTreated(boolean treated) { isTreated = treated; }
    
    public String getLocation() { return location; }
    public void setLocation(String location) { this.location = location; }
    
    public String getWeatherConditions() { return weatherConditions; }
    public void setWeatherConditions(String weatherConditions) { this.weatherConditions = weatherConditions; }
    
    public long getDiagnosisDate() { return diagnosisDate; }
    public void setDiagnosisDate(long diagnosisDate) { this.diagnosisDate = diagnosisDate; }
    
    public long getLastUpdated() { return lastUpdated; }
    public void setLastUpdated(long lastUpdated) { this.lastUpdated = lastUpdated; }
}

File: lib/features/diagnosis/domain/entities/diagnosis.dart
class Diagnosis {
  final int? diagnosisId;
  final int livestockId;
  final int diseaseId;
  final String imagePath;
  final String diagnosisResult;
  final double confidence;
  final String status;
  final String? notes;
  final String? veterinarianNotes;
  final String? treatmentPrescribed;
  final String? followUpDate;
  final bool isTreated;
  final String? location;
  final String? weatherConditions;
  final DateTime diagnosisDate;
  final DateTime lastUpdated;

  const Diagnosis({
    this.diagnosisId,
    required this.livestockId,
    required this.diseaseId,
    required this.imagePath,
    required this.diagnosisResult,
    required this.confidence,
    required this.status,
    this.notes,
    this.veterinarianNotes,
    this.treatmentPrescribed,
    this.followUpDate,
    this.isTreated = false,
    this.location,
    this.weatherConditions,
    required this.diagnosisDate,
    required this.lastUpdated,
  });

  factory Diagnosis.create({
    required int livestockId,
    required String imagePath,
    required String diagnosisResult,
    required double confidence,
    required String status,
  }) {
    final now = DateTime.now();
    return Diagnosis(
      livestockId: livestockId,
      diseaseId: 0, // Will be set based on diagnosis result
      imagePath: imagePath,
      diagnosisResult: diagnosisResult,
      confidence: confidence,
      status: status,
      diagnosisDate: now,
      lastUpdated: now,
    );
  }

  Diagnosis copyWith({
    int? diagnosisId,
    int? livestockId,
    int? diseaseId,
    String? imagePath,
    String? diagnosisResult,
    double? confidence,
    String? status,
    String? notes,
    String? veterinarianNotes,
    String? treatmentPrescribed,
    String? followUpDate,
    bool? isTreated,
    String? location,
    String? weatherConditions,
    DateTime? diagnosisDate,
    DateTime? lastUpdated,
  }) {
    return Diagnosis(
      diagnosisId: diagnosisId ?? this.diagnosisId,
      livestockId: livestockId ?? this.livestockId,
      diseaseId: diseaseId ?? this.diseaseId,
      imagePath: imagePath ?? this.imagePath,
      diagnosisResult: diagnosisResult ?? this.diagnosisResult,
      confidence: confidence ?? this.confidence,
      status: status ?? this.status,
      notes: notes ?? this.notes,
      veterinarianNotes: veterinarianNotes ?? this.veterinarianNotes,
      treatmentPrescribed: treatmentPrescribed ?? this.treatmentPrescribed,
      followUpDate: followUpDate ?? this.followUpDate,
      isTreated: isTreated ?? this.isTreated,
      location: location ?? this.location,
      weatherConditions: weatherConditions ?? this.weatherConditions,
      diagnosisDate: diagnosisDate ?? this.diagnosisDate,
      lastUpdated: lastUpdated ?? this.lastUpdated,
    );
  }

  Map<String, dynamic> toMap() {
    return {
      'diagnosisId': diagnosisId,
      'livestockId': livestockId,
      'diseaseId': diseaseId,
      'imagePath': imagePath,
      'diagnosisResult': diagnosisResult,
      'confidence': confidence,
      'status': status,
      'notes': notes,
      'veterinarianNotes': veterinarianNotes,
      'treatmentPrescribed': treatmentPrescribed,
      'followUpDate': followUpDate,
      'isTreated': isTreated ? 1 : 0,
      'location': location,
      'weatherConditions': weatherConditions,
      'diagnosisDate': diagnosisDate.millisecondsSinceEpoch,
      'lastUpdated': lastUpdated.millisecondsSinceEpoch,
    };
  }

  factory Diagnosis.fromMap(Map<String, dynamic> map) {
    return Diagnosis(
      diagnosisId: map['diagnosisId'],
      livestockId: map['livestockId'],
      diseaseId: map['diseaseId'],
      imagePath: map['imagePath'],
      diagnosisResult: map['diagnosisResult'],
      confidence: map['confidence']?.toDouble() ?? 0.0,
      status: map['status'],
      notes: map['notes'],
      veterinarianNotes: map['veterinarianNotes'],
      treatmentPrescribed: map['treatmentPrescribed'],
      followUpDate: map['followUpDate'],
      isTreated: map['isTreated'] == 1,
      location: map['location'],
      weatherConditions: map['weatherConditions'],
      diagnosisDate: DateTime.fromMillisecondsSinceEpoch(map['diagnosisDate']),
      lastUpdated: DateTime.fromMillisecondsSinceEpoch(map['lastUpdated']),
    );
  }

  @override
  String toString() {
    return 'Diagnosis(diagnosisId: $diagnosisId, livestockId: $livestockId, diseaseId: $diseaseId, imagePath: $imagePath, diagnosisResult: $diagnosisResult, confidence: $confidence, status: $status)';
  }

  @override
  bool operator ==(Object other) {
    if (identical(this, other)) return true;
    return other is Diagnosis &&
        other.diagnosisId == diagnosisId &&
        other.livestockId == livestockId &&
        other.diseaseId == diseaseId &&
        other.imagePath == imagePath &&
        other.diagnosisResult == diagnosisResult &&
        other.confidence == confidence &&
        other.status == status;
  }

  @override
  int get hashCode {
    return diagnosisId.hashCode ^
        livestockId.hashCode ^
        diseaseId.hashCode ^
        imagePath.hashCode ^
        diagnosisResult.hashCode ^
        confidence.hashCode ^
        status.hashCode;
  }
}
